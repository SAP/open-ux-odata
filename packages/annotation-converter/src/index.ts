export * from './converter';
export * from './utils';
export * from './writeback';
