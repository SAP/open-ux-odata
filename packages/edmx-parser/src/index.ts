export { parse } from './parser';
export { merge } from './merger';
