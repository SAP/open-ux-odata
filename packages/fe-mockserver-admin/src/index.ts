import type { IMockserverPlugin } from '@sap-ux/fe-mockserver-core';
import * as path from 'node:path';

const mockServerAdminConfig: IMockserverPlugin = {
    name: 'mockserver-admin',
    services: [
        {
            alias: 'AdminService',
            urlPath: '/admin',
            metadataPath: path.join(__dirname, 'metadata.xml'),
            mockdataPath: path.join(__dirname, 'mockdata'),
            generateMockData: false
        }
    ]
};

export = mockServerAdminConfig;
