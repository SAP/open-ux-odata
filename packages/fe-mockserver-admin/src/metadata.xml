<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
    <edmx:DataServices>
        <Schema Namespace="MockServerAdmin" xmlns="http://docs.oasis-open.org/odata/ns/edm">
            <EntityType Name="Service">
                <Key>
                    <PropertyRef Name="UrlPath"/>
                </Key>
                <Property Name="UrlPath" Type="Edm.String" Nullable="false" MaxLength="255"/>
                <Property Name="InternalName" Type="Edm.String" Nullable="true" MaxLength="100"/>
                <Property Name="EncodedPath" Type="Edm.String" Nullable="true" MaxLength="255"/>
                <Property Name="ContextIsolation" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="Debug" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="Watch" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="HasETag" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="RegisteredAt" Type="Edm.DateTimeOffset" Nullable="false"/>
            </EntityType>
            <EntityContainer Name="AdminContainer">
                <EntitySet Name="Services" EntityType="MockServerAdmin.Service"/>
            </EntityContainer>
        </Schema>
    </edmx:DataServices>
</edmx:Edmx>