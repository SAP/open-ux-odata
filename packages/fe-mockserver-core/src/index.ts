import type { MockserverConfiguration } from './api';
import Router from 'router';
import type { IRouter } from 'router';
import { createMockMiddleware } from './middleware';
import * as path from 'path';

export interface IFileLoader {
    loadFile(filePath: string): Promise<string>;
    exists(filePath: string): Promise<boolean>;
    loadJS(filePath: string): Promise<any>;
}
export interface IMetadataProcessor {
    loadMetadata(filePath: string): Promise<string>;
}
export * from './api';
export { MockDataContributor } from './mockdata/functionBasedMockData';

export default class FEMockserver {
    isReady: Promise<void>;
    private fileLoader: IFileLoader;
    private metadataProvider: IMetadataProcessor;
    private readonly mainRouter: IRouter;

    constructor(private configuration: MockserverConfiguration) {
        this.mainRouter = new Router();
        this.isReady = this.initialize();
    }

    private async initialize() {
        const FileLoaderClass =
            (this.configuration.fileLoader as any) || (await import('./plugins/fileSystemLoader')).default;
        this.fileLoader = new FileLoaderClass() as IFileLoader;

        const MetadataProviderClass = (
            await this.fileLoader.loadJS(
                this.configuration.metadataProcessor?.name || path.resolve(__dirname, './plugins/metadataProvider')
            )
        ).default;
        this.metadataProvider = new MetadataProviderClass(
            this.fileLoader,
            this.configuration.metadataProcessor?.options
        ) as IMetadataProcessor;
        await createMockMiddleware(this.configuration, this.mainRouter, this.fileLoader, this.metadataProvider);
    }

    getRouter() {
        return this.mainRouter;
    }
}
