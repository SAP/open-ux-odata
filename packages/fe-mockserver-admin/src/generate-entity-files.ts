import { convert } from '@sap-ux/annotation-converter';
import { parse } from '@sap-ux/edmx-parser';
import type {
    Action,
    ConvertedMetadata,
    EntitySet,
    EntityType,
    NavigationProperty,
    RawMetadata
} from '@sap-ux/vocabularies-types';
import * as fs from 'fs';
import * as path from 'path';

export async function generateEntityFiles(METADATA_PATH: string, OUTPUT_DIR: string) {
    try {
        console.log('Loading metadata.xml...');
        const metadataContent = fs.readFileSync(METADATA_PATH, 'utf8');

        console.log('Parsing EDMX...');
        const parsedMetadata: RawMetadata = parse(metadataContent);

        console.log('Converting metadata with annotation converter...');
        const convertedMetadata: ConvertedMetadata = convert(parsedMetadata);

        // Generate TS files for each entity set
        for (const entitySet of convertedMetadata.entitySets) {
            console.log(`Generating TS file for ${entitySet.name}...`);
            await generateEntityFile(entitySet, convertedMetadata, OUTPUT_DIR);
        }

        await generateEntityContainerFile(convertedMetadata, OUTPUT_DIR);

        console.log('All entity files generated successfully!');
    } catch (error) {
        console.error('Error generating entity files:', error);
        process.exit(1);
    }
}

async function generateEntityContainerFile(convertedMetadata: ConvertedMetadata, OUTPUT_DIR: string): Promise<void> {
    const actions = convertedMetadata.actionImports;
    if (actions.length > 0) {
        const actionSwitchCase = actions
            .map((action) => {
                return `case '${action.name}':
            break; // TODO: Implement ${action.name} action`;
            })
            .join('\n');

        const mockEntityContainer = `import type {MockEntityContainerContributor, ODataRequest, Action} from "@sap-ux/ui5-middleware-fe-mockserver";


const EntityContainer: MockEntityContainerContributor = {
    async executeAction(actionDefinition: Action, _actionData: any, _keys: Record<string, unknown>, _odataRequest: ODataRequest): Promise<unknown> {
        switch(actionDefinition.name) {
            ${actionSwitchCase}
            // Implement action handlers here
            default:
                return undefined; // No custom action handler, proceed with default handling
        }
    }
}
export default EntityContainer;`;
        const outputPath = path.join(OUTPUT_DIR, `EntityContainer.ts`);
        fs.writeFileSync(outputPath, mockEntityContainer);
    }
}

async function generateEntityFile(
    entitySet: EntitySet,
    metadata: ConvertedMetadata,
    OUTPUT_DIR: string
): Promise<void> {
    const entityType = entitySet.entityType;
    const keyName = entityType.name.split('.').pop() + 'Keys';
    const actionDataName = entityType.name.split('.').pop() + 'ActionData';
    // Get actions for this entity set
    const actions = entitySet.entityType.actions;

    // Get navigation properties without constraints for this entity set
    const entityNavPropsWithoutConstraints = entityType.navigationProperties.filter(
        (prop) =>
            prop.name !== 'DraftAdministrativeData' &&
            prop.name !== 'SiblingEntity' && // Exclude DraftAdministrativeData and SiblingEntity
            prop.referentialConstraint.length === 0 &&
            (!prop.partner ||
                prop.targetType.navigationProperties.by_name(prop.partner)?.referentialConstraint.length === 0)
    );

    // Generate TypeScript interface for the entity
    const entityInterface = generateEntityInterface(entityType);
    const entityNavPropNamesInterface = generateEntityNavPropNamesInterface(entityType);
    const entityNavPropTypesInterface = generateEntityNavPropTypesInterface(entityType);
    const entityActionParams = generateEntityActionParams(entitySet, actions);
    // Generate action switch cases
    const actionSwitchCases = generateActionSwitchCases(actions);
    let executeActionMethod = '';
    if (Object.keys(actions).length > 0) {
        executeActionMethod = `async executeAction(actionDefinition: Action, actionData: ${actionDataName}, _keys: ${keyName}, _odataRequest: ODataRequest): Promise<object | undefined> {
        switch(actionData._type) {
${actionSwitchCases}
            default:
                console.warn(\`Unhandled action: \${actionDefinition.name}\`);
                return undefined;
        }
    }`;
    }

    // Generate getReferentialConstraints method
    const getReferentialConstraintsMethod = generateGetReferentialConstraintsMethod(entityNavPropsWithoutConstraints);
    const allMethods = [];
    if (executeActionMethod.length) {
        allMethods.push(executeActionMethod);
    }
    if (getReferentialConstraintsMethod.length) {
        allMethods.push(getReferentialConstraintsMethod);
    }
    const fileContent = `import type { MockDataContributor, ODataRequest, Action, NavigationProperty, PartialReferentialConstraint } from "@sap-ux/ui5-middleware-fe-mockserver"
${entityNavPropTypesInterface}
${entityNavPropNamesInterface}
${entityInterface}
${entityActionParams}


const ${entitySet.name}: MockDataContributor<${entitySet.entityTypeName.split('.').pop()}Type> = {
    ${allMethods.join(',')}
}

export default ${entitySet.name};
`;

    //if(allMethods.length > 0) {
    const outputPath = path.join(OUTPUT_DIR, `${entitySet.name}.ts`);
    fs.writeFileSync(outputPath, fileContent);
    console.log(`Generated: ${outputPath}`);
    //} else {
    //   console.log(`No method to generate for ${entitySet.name}, skipping file creation.`);
    //}
}

function generateEntityNavPropNamesInterface(entityType: EntityType): string {
    const typeName = entityType.name.split('.').pop() + 'NavPropNames';

    const properties: string[] = [];

    for (const prop of entityType.navigationProperties) {
        if (prop.name !== 'DraftAdministrativeData') {
            properties.push(`"${prop.name}"`);
        }
    }
    if (!entityType || !entityType.navigationProperties || properties.length === 0) {
        return `export type ${typeName} = {
    // Navigation properties not available in metadata
}`;
    }
    return `export type ${typeName} = ${properties.join(' | ')};`;
}
function generateEntityNavPropTypesInterface(entityType: EntityType): string {
    const currentType = entityType.name.split('.').pop() + 'Type';
    const typeName = entityType.name.split('.').pop() + 'NavPropTypes';
    const importStatements: Set<string> = new Set();
    if (!entityType || !entityType.navigationProperties) {
        return `export type ${typeName} = {
    // Navigation properties not available in metadata
}`;
    }

    const properties: string[] = [];

    for (const prop of entityType.navigationProperties) {
        const targetTypeName = prop.targetType.name.split('.').pop() + 'Type';
        if (targetTypeName !== 'DraftAdministrativeDataType' && targetTypeName !== currentType) {
            importStatements.add(
                `import type { ${targetTypeName} } from "./${prop.targetType.name.split('.').pop()}";`
            );
        }
        if (targetTypeName !== 'DraftAdministrativeDataType') {
            if (prop.isCollection) {
                properties.push(`${prop.name}: ${targetTypeName}[];`);
            } else {
                properties.push(`${prop.name}: ${targetTypeName};`);
            }
        }
    }

    return `${Array.from(importStatements).join('\n')}
export type ${typeName} = {
${properties.join('\n')}
}`;
}
function generateEntityInterface(entityType: EntityType): string {
    const typeName = entityType.name.split('.').pop() + 'Type';
    const keyName = entityType.name.split('.').pop() + 'Keys';

    if (!entityType || !entityType.keys) {
        return `export type ${typeName} = {
    // Entity type definition not available in metadata
};
export type ${keyName} = {
    // Entity type definition not available in metadata
};`;
    }

    const keys: string[] = [];
    const properties: string[] = [];

    // Add key properties
    if (entityType.keys) {
        for (const key of entityType.keys) {
            const prop = entityType.entityProperties.by_name(key.name);
            if (prop) {
                const tsType = mapODataTypeToTypeScript(prop.type);
                const nullable = prop.nullable ? '?' : '';
                properties.push(`    ${key.name}${nullable}: ${tsType};`);
                keys.push(`    ${key.name}${nullable}: ${tsType};`);
            }
        }
    }

    // Add other entity properties
    if (entityType.entityProperties) {
        for (const prop of entityType.entityProperties) {
            if (!entityType.keys?.some((k: any) => k.name === prop.name)) {
                const tsType = mapODataTypeToTypeScript((prop as any).type);
                const nullable = (prop as any).nullable ? '?' : '';
                properties.push(`    ${prop.name}${nullable}: ${tsType};`);
            }
        }
    }

    return `export type ${typeName} = {
${properties.join('\n')}
};

export type ${keyName} = {
${keys.join('\n')}
};`;
}

function mapODataTypeToTypeScript(odataType: string): string {
    const typeMap: Record<string, string> = {
        'Edm.String': 'string',
        'Edm.Int32': 'number',
        'Edm.Int64': 'number',
        'Edm.Double': 'number',
        'Edm.Decimal': 'number',
        'Edm.Boolean': 'boolean',
        'Edm.DateTime': 'Date',
        'Edm.DateTimeOffset': 'Date',
        'Edm.Date': 'Date',
        'Edm.Time': 'Date',
        'Edm.Guid': 'string',
        'Edm.Binary': 'string'
    };

    return typeMap[odataType] || 'any';
}

function generateEntityActionParams(entitySet: EntitySet, actions: Record<string, Action>): string {
    if (Object.keys(actions).length === 0) {
        return '';
    }

    const actionParams = Object.keys(actions)
        .filter((actionName) => {
            return (
                !actionName.endsWith('draftEdit') &&
                !actionName.endsWith('draftActivate') &&
                !actionName.endsWith('draftPrepare') &&
                !actionName.endsWith('draftDiscard')
            );
        })
        .map((actionName) => {
            const action = actions[actionName];
            if (action.parameters && action.parameters.length > 0) {
                const params = action.parameters
                    .map((param) => {
                        let tsType = mapODataTypeToTypeScript(param.type);
                        if (tsType === 'any' && param.type === entitySet.entityTypeName) {
                            tsType = entitySet.entityTypeName.split('.').pop() + 'Type';
                        }
                        const optional = param.nullable ? '?' : '';
                        return `    ${param.name}${optional}: ${tsType};`;
                    })
                    .join('\n');
                return `    | {\n    _type: '${action.name}';\n${params}\n}`;
            }
        })
        .join('\n');

    return `export type ${entitySet.entityType.name}ActionData =\n${actionParams};`;
}

function generateActionSwitchCases(actions: Record<string, Action>): string {
    return Object.keys(actions)
        .filter(
            (actionName) =>
                !actionName.endsWith('draftEdit') &&
                !actionName.endsWith('draftActivate') &&
                !actionName.endsWith('draftPrepare') &&
                !actionName.endsWith('draftDiscard')
        )
        .map(
            (actionName) => `            case '${actions[actionName].name}':
                // TODO: Implement ${actions[actionName].name} action
                console.log('Executing action: ${actions[actionName].name}');
                return {};`
        )
        .join('\n');
}

function generateGetReferentialConstraintsMethod(navPropsWithoutConstraints: NavigationProperty[]): string {
    if (navPropsWithoutConstraints.length === 0) {
        return '';
    }

    const allPotentialCases = navPropsWithoutConstraints
        .map(
            (prop) =>
                `
            case '${prop.name}':  // Navigation to ${prop.targetTypeName}
                return []; // TODO add the missing referential constraints
            break;`
        )
        .join('\n');

    return `
    getReferentialConstraints: (navigationProperty: NavigationProperty):PartialReferentialConstraint[] | undefined =>  {
        // Navigation properties without referential constraints:
        switch(navigationProperty.name) {
             ${allPotentialCases}
             default:
                return navigationProperty.referentialConstraint;
        }
    }`;
}
