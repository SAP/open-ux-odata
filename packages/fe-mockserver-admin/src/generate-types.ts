import { convert } from '@sap-ux/annotation-converter';
import { parse } from '@sap-ux/edmx-parser';
import type {
    ComplexType,
    ConvertedMetadata,
    EntityType,
    RawMetadata,
    TypeDefinition
} from '@sap-ux/vocabularies-types';
import fs from 'fs';

// Helper to build lookup maps for types
function buildTypeMaps(metadata: ConvertedMetadata) {
    const entityTypeMap = new Map<string, any>();
    const complexTypeMap = new Map<string, any>();
    const typeDefMap = new Map<string, any>();

    for (const et of metadata.entityTypes) {
        entityTypeMap.set(et.fullyQualifiedName || et.name, et);
    }
    for (const ct of metadata.complexTypes) {
        complexTypeMap.set(ct.fullyQualifiedName || ct.name, ct);
    }
    for (const td of metadata.typeDefinitions) {
        typeDefMap.set(td.fullyQualifiedName || td.name, td);
    }
    return { entityTypeMap, complexTypeMap, typeDefMap };
}

// Enhanced mapping function to handle complex types and typeDefs
function mapEdmTypeToTsType(
    edmType: string,
    typeMaps: { entityTypeMap: Map<string, any>; complexTypeMap: Map<string, any>; typeDefMap: Map<string, any> }
): string {
    // Handle collection types
    const collectionMatch = edmType.match(/^Collection\((.+)\)$/);
    if (collectionMatch) {
        const innerType = mapEdmTypeToTsType(collectionMatch[1], typeMaps);
        return `${innerType}[]`;
    }

    // Primitive EDM types
    switch (edmType) {
        case 'Edm.String':
            return 'string';
        case 'Edm.Boolean':
            return 'boolean';
        case 'Edm.Int32':
        case 'Edm.Int16':
        case 'Edm.Int64':
        case 'Edm.Decimal':
        case 'Edm.Double':
        case 'Edm.Single':
        case 'Edm.Byte':
        case 'Edm.SByte':
            return 'number';
        case 'Edm.DateTimeOffset':
        case 'Edm.Date':
        case 'Edm.TimeOfDay':
            return 'string';
        case 'Edm.Guid':
        case 'Edm.Binary':
            return 'string';
    }

    // Complex types
    if (typeMaps.complexTypeMap.has(edmType)) {
        return typeMaps.complexTypeMap.get(edmType).name;
    }
    // Type definitions
    if (typeMaps.typeDefMap.has(edmType)) {
        return typeMaps.typeDefMap.get(edmType).name;
    }
    // Entity types (rare for property, but possible for navigation)
    if (typeMaps.entityTypeMap.has(edmType)) {
        return typeMaps.entityTypeMap.get(edmType).name;
    }

    // Fallback: strip namespace if present
    return edmType.split('.').pop() || edmType;
}

// Generate TypeScript type for an entity type and its keys
function generateEntityType(entityType: EntityType, typeMaps: any): string {
    let result = `export type ${entityType.name} = {\n`;
    for (const prop of entityType.entityProperties) {
        const tsType = mapEdmTypeToTsType(prop.type, typeMaps);
        result += `  ${prop.name}${prop.nullable ? '?' : ''}: ${tsType};\n`;
    }

    // Include navigation properties (wrapped with NavPropTo<...>)
    if (entityType.navigationProperties && entityType.navigationProperties.length > 0) {
        for (const nav of entityType.navigationProperties) {
            // Resolve target entity type name
            let targetName: string;
            if ((nav as any).targetType && (nav as any).targetType.name) {
                targetName = (nav as any).targetType.name;
            } else {
                const fqn = (nav as any).targetTypeName || (nav as any).targetType?.fullyQualifiedName || '';
                const et = typeMaps.entityTypeMap.get(fqn) || typeMaps.entityTypeMap.get((fqn || '').toString());
                targetName = et?.name || (fqn ? fqn.split('.').pop() : 'any');
            }

            const navInnerType = nav.isCollection ? `${targetName}[]` : targetName;
            // Navigation properties are typically optional in generated types
            result += `  ${nav.name}?: NavPropTo<${navInnerType}>;\n`;
        }
    }

    result += '}\n';

    result += `export type ${entityType.name}Keys = {\n`;
    for (const prop of entityType.keys) {
        const tsType = mapEdmTypeToTsType(prop.type, typeMaps);
        result += `  ${prop.name}${prop.nullable ? '?' : ''}: ${tsType};\n`;
    }
    result += '}\n';

    return result;
}

// Generate TypeScript type for a complex type
function generateComplexType(complexType: ComplexType, typeMaps: any): string {
    let result = `export type ${complexType.name} = {\n`;
    for (const prop of complexType.properties) {
        const tsType = mapEdmTypeToTsType(prop.type, typeMaps);
        result += `  ${prop.name}${prop.nullable ? '?' : ''}: ${tsType};\n`;
    }
    result += '}\n';
    return result;
}

// Generate TypeScript type alias for a type definition
function generateTypeDefinition(typeDefinition: TypeDefinition, typeMaps: any): string {
    const tsType = mapEdmTypeToTsType(typeDefinition.underlyingType, typeMaps);
    return `export type ${typeDefinition.name} = ${tsType};\n`;
}

export async function generateTypes(METADATA_PATH: string, OUTPUT_DIR: string) {
    try {
        console.log('Loading metadata.xml...');
        const metadataContent = fs.readFileSync(METADATA_PATH, 'utf8');

        console.log('Parsing EDMX...');
        const parsedMetadata: RawMetadata = parse(metadataContent);

        console.log('Converting metadata with annotation converter...');
        const convertedMetadata: ConvertedMetadata = convert(parsedMetadata);

        // Build type maps for reference resolution
        const typeMaps = buildTypeMaps(convertedMetadata);

        let typeDefinitions = 'export type NavPropTo<T> = T;\n\n';
        // Generate TS files for each entity set
        typeDefinitions += '// ====== Entity Types ======\n\n';
        for (const entityType of convertedMetadata.entityTypes) {
            console.log(`Generating TS file for ${entityType.name}...`);
            typeDefinitions += '// Type definitions for ' + entityType.name + '\n';
            typeDefinitions += generateEntityType(entityType, typeMaps);
            typeDefinitions += '\n';
        }

        if (convertedMetadata.complexTypes.length > 0) {
            typeDefinitions += '// ====== Complex Types ======\n\n';
            for (const complexType of convertedMetadata.complexTypes) {
                console.log(`Generating TS file for ${complexType.name}...`);
                typeDefinitions += '// Type definitions for ' + complexType.name + '\n';
                typeDefinitions += generateComplexType(complexType, typeMaps);
                typeDefinitions += '\n';
            }
        }

        if (convertedMetadata.typeDefinitions.length > 0) {
            typeDefinitions += '// ====== Type Definition ===\n\n';
            for (const typeDefinition of convertedMetadata.typeDefinitions) {
                console.log(`Generating TS file for ${typeDefinition.name}...`);
                typeDefinitions += '// Type definitions for ' + typeDefinition.name + '\n';
                typeDefinitions += generateTypeDefinition(typeDefinition, typeMaps);
                typeDefinitions += '\n';
            }
        }

        const outputPath = `${OUTPUT_DIR}/ODataTypes.d.ts`;
        fs.writeFileSync(outputPath, typeDefinitions, 'utf8');
        console.log(`Type definitions written to ${outputPath}`);
    } catch (error) {
        console.error('Error generating entity files:', error);
        process.exit(1);
    }
}
