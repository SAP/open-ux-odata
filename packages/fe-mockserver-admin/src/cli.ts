#!/usr/bin/env node

import { Command } from 'commander';
import * as fs from 'fs';
import * as path from 'path';
import { generateEntityFiles } from './generate-entity-files';

const program = new Command();

program.name('fe-mockserver-admin').description('CLI for FE Mockserver administration tasks').version('0.0.2');

program
    .command('generate-types')
    .description('Generate TypeScript types from OData metadata')
    .requiredOption('-m, --metadata <path>', 'Path to the metadata.xml file')
    .option('-o, --output <path>', 'Output directory for generated types (defaults to metadata directory)')
    .action(async (options) => {
        try {
            const metadataPath = path.resolve(options.metadata);
            const outputDir = options.output ? path.resolve(options.output) : path.dirname(metadataPath);

            // Validate metadata file exists

            if (!fs.existsSync(metadataPath)) {
                console.error(`Error: Metadata file not found at: ${metadataPath}`);
                process.exit(1);
            }

            // Ensure output directory exists
            if (!fs.existsSync(outputDir)) {
                fs.mkdirSync(outputDir, { recursive: true });
            }
            console.log(`Loading metadata from: ${metadataPath}`);
            console.log(`Output directory: ${outputDir}`);

            const { generateTypes } = await import('./generate-types');
            await generateTypes(metadataPath, outputDir);
        } catch (error) {
            console.error('Error generating types:', error);
            process.exit(1);
        }
    });

program
    .command('generate-entity-files')
    .description('Generate entity files from OData metadata')
    .requiredOption('-m, --metadata <path>', 'Path to the metadata.xml file')
    .option('-o, --output <path>', 'Output directory for generated files (defaults to metadata directory)')
    .action(async (options) => {
        try {
            const metadataPath = path.resolve(options.metadata);
            const outputDir = options.output ? path.resolve(options.output) : path.dirname(metadataPath);

            // Validate metadata file exists
            if (!fs.existsSync(metadataPath)) {
                console.error(`Error: Metadata file not found at: ${metadataPath}`);
                process.exit(1);
            }

            // Ensure output directory exists
            if (!fs.existsSync(outputDir)) {
                fs.mkdirSync(outputDir, { recursive: true });
            }

            console.log(`Loading metadata from: ${metadataPath}`);
            console.log(`Output directory: ${outputDir}`);

            await generateEntityFiles(metadataPath, outputDir);

            console.log('All entity files generated successfully!');
        } catch (error) {
            console.error('Error generating entity files:', error);
            process.exit(1);
        }
    });

// Parse arguments and execute
program.parse();
