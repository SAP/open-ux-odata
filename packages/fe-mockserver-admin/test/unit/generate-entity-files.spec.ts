import * as fs from 'fs';
import * as path from 'path';
import { generateEntityFiles } from '../../src/generate-entity-files';

describe('generate-entity-files with real metadata', () => {
    const testDataDir = path.join(__dirname, '__testData');
    const metadataPath = path.join(testDataDir, 'metadata.xml');
    const outputDir = path.join(testDataDir, 'generated');

    beforeAll(() => {
        // Clean up generated files
        if (fs.existsSync(outputDir)) {
            const files = fs.readdirSync(outputDir);
            for (const file of files) {
                fs.unlinkSync(path.join(outputDir, file));
            }
            fs.rmdirSync(outputDir);
        }
        // Ensure output directory exists
        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir, { recursive: true });
        }
    });

    afterAll(() => {
        // Clean up generated files
        if (fs.existsSync(outputDir)) {
            const files = fs.readdirSync(outputDir);
            for (const file of files) {
                fs.unlinkSync(path.join(outputDir, file));
            }
            fs.rmdirSync(outputDir);
        }
    });

    it('should generate entity files from real metadata.xml', async () => {
        // Verify metadata file exists
        expect(fs.existsSync(metadataPath)).toBe(true);

        // Generate entity files
        await generateEntityFiles(metadataPath, outputDir);

        // Check that files were generated
        const generatedFiles = fs.readdirSync(outputDir);
        expect(generatedFiles.length).toBeGreaterThan(0);

        // Should generate RootEntity.ts based on the metadata
        const rootEntityFile = path.join(outputDir, 'RootEntity.ts');
        expect(fs.existsSync(rootEntityFile)).toBe(true);

        // Read and validate the generated RootEntity.ts content
        const rootEntityContent = fs.readFileSync(rootEntityFile, 'utf8');
        expect(rootEntityContent).toMatchSnapshot();

        // Log the generated content for manual inspection
        console.log('Generated RootEntity.ts content:');
        console.log('='.repeat(50));
        console.log(rootEntityContent);
        console.log('='.repeat(50));
    });

    it('should generate EntityContainer.ts if action imports exist', async () => {
        // Generate entity files
        await generateEntityFiles(metadataPath, outputDir);

        // Check if EntityContainer.ts was generated (depends on metadata having action imports)
        const entityContainerFile = path.join(outputDir, 'EntityContainer.ts');

        if (fs.existsSync(entityContainerFile)) {
            const entityContainerContent = fs.readFileSync(entityContainerFile, 'utf8');
            expect(entityContainerContent).toMatchSnapshot();
        } else {
            console.log('No EntityContainer.ts generated (no action imports in metadata)');
        }
    });

    it('should validate metadata parsing and conversion', async () => {
        // This test validates the parsing process
        const metadataContent = fs.readFileSync(metadataPath, 'utf8');

        // Basic validation that metadata contains expected elements
        expect(metadataContent).toContain('EntitySet Name="RootEntity"');
        expect(metadataContent).toContain('EntityType Name="RootEntity"');
        expect(metadataContent).toContain('Action Name="myCustomAction"');
        expect(metadataContent).toContain('NavigationProperty Name="SiblingEntity"');

        // Generate files and verify no errors
        await expect(generateEntityFiles(metadataPath, outputDir)).resolves.not.toThrow();
    });
});
