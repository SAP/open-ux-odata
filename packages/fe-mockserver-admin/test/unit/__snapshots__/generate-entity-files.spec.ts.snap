// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-entity-files with real metadata should generate entity files from real metadata.xml 1`] = `
"import type { ODataRequest, Action, NavigationProperty, PartialReferentialConstraint } from "@sap-ux/ui5-middleware-fe-mockserver"
import { MockDataContributorClass } from "@sap-ux/ui5-middleware-fe-mockserver"
import type { RootEntity, RootEntityKeys} from "./ODataTypes";
type RootEntityAction_myCustomActionData = {
    in?: RootEntity;
}
;export type RootEntityActionData =
    | {
    _type: 'myCustomAction';
} & RootEntityAction_myCustomActionData;


export default class RootEntityContributor extends MockDataContributorClass<RootEntity> {
        async myCustomAction(_actionDefinition: Action, _actionData: RootEntityAction_myCustomActionData, _keys: RootEntityKeys, _odataRequest: ODataRequest): Promise<unknown> { 
                    throw new Error('myCustomAction is not implemented');
                }

async executeAction(actionDefinition: Action, actionData: RootEntityActionData, keys: RootEntityKeys, odataRequest: ODataRequest): Promise<object | undefined> {
        switch(actionData._type) {
            case 'myCustomAction':
                  return this.myCustomAction(actionDefinition, actionData, keys, odataRequest);
            default:
                this.throwError(\`Action \${actionDefinition.name} not implemented\`, 501);
                return;
        }
    }
}
"
`;
